### PR TITLE
Refactor mods to support more general data types

### PR DESCRIPTION
**Pull request by GiggleLiu**
Nice package, I did some changes to fit my using case. List of changes are,
* add a second type parameter to Mod, so that `UInt` can be used in computing, `GaussMod{N,T}` is now an alias of `Mod{N,Complex{T}}`, so that a lot code can be reused.
* improve the performance of `Mod` algebra a lot.
* Clean up codes
    * use promotion instead of redefining algebras with integers and rationals.
* make CRT compatible with big integers, so that one does not suffer from integer overflow when using CRT to compute extremely big numbers. One the otherside, it will not return the modulus anymore, because the modulus can sometimes be a BigInt, which is not a valid type parameter.
* update documentations and README.

Looking forward to your review.

### CODE REVIEW

**Comment by GiggleLiu** [comment]
I notice there is a better way to implement this. Let me polish it a bit before you can review  :P

**Comment by GiggleLiu** [comment]
Now I think this PR is ready for review @scheinerman . I would suggest reading the tests so that you can get what I have changed quickly. Hope these changes does not break your projects built on top of it.

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl]: This calls `mod(_, N)` twice. Wouldn't be like `iszero(x - y)` more efficient?

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl:4]: Where are methods for `zero`, `one` defined now?
Are `iszero` and `oneunit` still missing?

**Comment by scheinerman** [comment]
@GiggleLiu You did an amazing amount of work on this. Thank you.

**Comment by scheinerman** [comment]
@GiggleLiu When I run the checks the documentation build doesn't work. I've never built documentation (I just use the `README` file) so I don't know what to do. Would you like to see what the problem is before I merge and release?

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl]: that is broken, if `x.val === typemin(T)`.

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl:117]: Why can't we use `invmod` in `Signed` case like in `Unsigned` case ?

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl]: Where is this implemented. (`x'` compiles to `adjoint(x)`, which is not `inv(x)` ).

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl:177]: unhandled integer overflow, if type of `primes` is not `BigInt`. In the typical use case `primes` are small, but result is big.
Not sure if this algorithm is efficient, because all arithmetic need to be done in maximal size integers (of type `BigInt`) to be correct.

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl]: see above: result type maybe frequently too small. Better user `BigInt` as result type.

Coresponding test cases are missing (for the overflowing case).

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl]: Type stability issue: The returned `1` has not always the same type as the final return value (`typeof(M)`).

File: test/runtests.jl
Comment (KlausC) [inline test/runtests.jl:45]: Why do you ban `GaussMod` from the tests, it is stell defined and to be tested.

File: src/GaussMods.jl
Comment (KlausC) [inline src/GaussMods.jl:15]: missing test case.

**Review by KlausC — CHANGES_REQUESTED** [review]
Good improvements.
I would generally add a test coverage report, because it is not clear, if all branches of the code are proved by test cases.

**Review by scheinerman — APPROVED** [review]
I am a GitHub noob, so I'm not sure what I need to do. Is this ready for me to "merge" and how do I do that? Has the documentation error been fixed?

**Comment by KlausC** [comment]
@scheinerman if you think, the changes are all ok, you can press the "merge" button; that will insert all changes into the main development branch.

Neverthless there are quite some undiscussed review comments, which should be resolved by the author of the PR, before you approve and merge the changes. You can add your comments to the individual review threads.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl:117]: Good point, `gcdx` is faster, but not applicable for `UInt`. Using `invmod` in both cases will cause regression in the performance of `Int` type.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl]: Do you mean making BigInt as the default?

File: src/GaussMods.jl
Comment (GiggleLiu) [reply src/GaussMods.jl:15]: Added the test. Wondering where did you see the test coverage? Can I change the travis CI to github actions and add the code coverage in this PR so that I can check the coverage more carefully?

TYI: Travis CI is becoming very restrictive on the amount of jobs one can submit, so people are switching to Github actions.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl]: You are right, I should define iszero instead of defining `==`.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl]: I throwed an error instead. It does not make sense to feed an empty vector into this function. Good?

File: test/runtests.jl
Comment (GiggleLiu) [reply test/runtests.jl:45]: I added this `GaussMod` back to make sure this PR is compatible with previous interface.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl]: Right, I removed this comment. For numbers, `adjoint `is same as `conj`. I see `conj` already defined in the file.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl]: Good catch, fixed with a test case.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl:4]: They are automatically generated. Since Mod subtypes Number, it will use `Mod{N,T}(0)` as the zero element and `Mod{N,T}(1)` as the one and oneunit. `iszero` can be defined by `== 0`. Since iszero is more efficient, I switch to defining `iszero` first and use `iszero` for defining `==` in the new commits.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl:177]: This function does not change the element types automatically. If one wants to compute with BigInt, he should input BigInt vectors. Users are supposed to use other interfaces bellow.

**Comment by GiggleLiu** [comment]
Hi, @scheinerman . I have fixed the documentation, the error was caused by the removal of docstrings of `GaussMod`, now I have added it back. Cheers.

And here is a not so important tip: After @KlausC approves the changes, you can merge the this PR. There is a dropdown menu along side the merge button, among different option to merge a PR in the dropdown menu, the "Squash and merge" option is recommended because it will keep the commit history clean.

**Review by scheinerman — COMMENTED** [review]
@GiggleLiu Is this ready to be merged?

**Comment by GiggleLiu** [comment]
@scheinerman I am still waiting for the feedback of @KlausC . We will ping you after all issues being resolved.

**Comment by scheinerman** [comment]
> @scheinerman I am still waiting for the feedback of @KlausC . We will ping you after all issues being resolved.

Thanks @GiggleLiu . (This is all so confusing to me!)

File: src/GaussMods.jl
Comment (KlausC) [reply src/GaussMods.jl:15]: I did not "see" the missing test coverage, but did not find a test case in the test files, and I did not find a link to test coverage results in the README.

If you want to switch to github - CI, I give you a link to my current package; see directores `.ci`, `.github`, `.codecov.yml` and `README.md`. Most of the files in there have to be customized - but that is easy.

Probably the github actions have to be activated by the repository owner.

File: src/Mods.jl
Comment (KlausC) [reply src/Mods.jl:4]: ok.

File: src/Mods.jl
Comment (KlausC) [reply src/Mods.jl:117]: I see. `invmod` has to check special cases wit negative arguments and performs a `mod`, which are not necessary in this use case. Maybe write a comment to avoid this question in the future,

File: src/Mods.jl
Comment (KlausC) [reply src/Mods.jl:177]: Yes. But then a check for overflow of `M` is needed in order to not obtain false results without warning.

File: src/Mods.jl
Comment (KlausC) [reply src/Mods.jl]: That was my idea, but otherwise see above (check overflow).

File: test/runtests.jl
Comment (KlausC) [reply test/runtests.jl:45]: It is required in the tests because it is defined in the src.

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl]: A sensible return value would be `zero(eltype(primes))`.

**Review by KlausC — COMMENTED** [review]
Still a few open issues.
Switching CI and test coverage is maybe worth a separate PR.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl]: It is not intuitive to me whether it should return zero or one for empty inputs. Do you have any arguments favoring zero?

File: test/runtests.jl
Comment (GiggleLiu) [reply test/runtests.jl:45]: Yeah, I meant I added GaussMod back to the tests.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl:177]: This makes sense. I changed the default type to BigInt.
If a user want to use Int64 for computing, he should write `CRT(Int64, args...)`
There will be no overflow error, because the user is aware of issue by writting this explicitly.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl:117]: Sounds good, added some comments.

File: src/GaussMods.jl
Comment (GiggleLiu) [reply src/GaussMods.jl:15]: Yeah, let's fix the CI in a seperate PR. Added the missing test case.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl]: BTW, there is another reason why I choose to throw an error. The `eltype` for a tuple of length 0 is ill-defined, while this interfaces aim to support both vectors and tuples.

File: src/Mods.jl
Comment (KlausC) [inline src/Mods.jl]: There is no default `BigInt` argument, opposed to the note.

File: src/Mods.jl
Comment (KlausC) [reply src/Mods.jl]: If you specify the result type as the first argument (which defaults to `BigInt`) as you suppose, the type of the return value is clear.
Arguments for `0` as result for empty inputs are for example:

1. the condition `0 <= crt(r, p) < prod(p)` for empty `p` and `prod([]) == 1`
2. for integer arrays, the rest of the algorithm executes and delivers `0`.
3. it is a corner case in the identity `crt([r1; r2], [p1;p2]) == crt(crt(r1, p1), crt(r2, p2)], [lcm(p1); lcm(p2)])`

File: src/GaussMods.jl
Comment (KlausC) [reply src/GaussMods.jl:15]: BTW: If you run the tests locally, ther is a coverage option:
```
]activate Mods
test --coverage
```
which leaves `*.jl.cov` files in `src`.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl]: Thanks for detailed explaination, it makes sense. Return 0 instead in the new commit.

File: src/Mods.jl
Comment (GiggleLiu) [reply src/Mods.jl]: Added the type argument in the new commit.

**Review by KlausC — APPROVED** [review]
That was a good job! I am happy with the changes now.

**Comment by scheinerman** [comment]
Dear @GiggleLiu  and @KlausC : I appreciate your interest and work on this `Mods` module. Is it ready to be merged??? (I'll be offline all day today -- Sunday -- so I will see your answers tomorrow.)

**Comment by GiggleLiu** [comment]
@scheinerman Yes, it is ready to be merged. Please go ahead.
BTW, may I ask for a new release like 1.3.0 after merging this pr?

(No worries, have a nice weekend!)

**Comment by scheinerman** [comment]
> @scheinerman Yes, it is ready to be merged. Please go ahead.
> BTW, may I ask for a new release like 1.3.0 after merging this pr?
> 
> (No worries, have a nice weekend!)

Yes. I'll test it out and then publish a release through JuliaHub at 1.3.0 as you suggested. There are enough changes that we should increase the middle number. Thank you again for your contributions!

**Comment by scheinerman** [comment]
@GiggleLiu @KlausC @charleskawczynski  : I see that the `README` says that docs are passing, but where do I find the docs? (I've always just used the `README` to document modules.)

**Comment by GiggleLiu** [comment]
> I see that the README says that docs are passing, but where do I find the docs?

I think there are some issue related to deployment, has the documentation been successfully deployed before? I can not find any existing version.

The current error message is 

```
 ! [remote rejected] HEAD -> gh-pages (cannot lock ref 'refs/heads/gh-pages': is at f96270b502aeea926c927aa672e5defe9506a77a but expected 7abb1c83c1bde0c0f81a83b07abbcf30f7e04028)
106
error: failed to push some refs to 'https://github.com/scheinerman/Mods.jl.git'
107
┌ Error: Failed to push:
108
│   exception =
109
│    failed process: Process(`git push -fq upstream HEAD:gh-pages`, ProcessExited(1)) [1]
110
│    
111
│    Stacktrace:
112
│      [1] pipeline_error
```
It is a git push command error, maybe a permission issue? I am not sure.


> (I've always just used the README to document modules.)

Me too. Actually this is a very good practice to host documents in README as long as a single page can do the job. Without documentation, one can still check the APIs in an REPL. So, no worry if the documentation issue can not be fixed.

**Comment by scheinerman** [comment]
> > I see that the README says that docs are passing, but where do I find the docs?
> 
> I think there are some issue related to deployment, has the documentation been successfully deployed before? I can not find any existing version.
> 
> The current error message is
> 
> ```
>  ! [remote rejected] HEAD -> gh-pages (cannot lock ref 'refs/heads/gh-pages': is at f96270b502aeea926c927aa672e5defe9506a77a but expected 7abb1c83c1bde0c0f81a83b07abbcf30f7e04028)
> 106
> error: failed to push some refs to 'https://github.com/scheinerman/Mods.jl.git'
> 107
> ┌ Error: Failed to push:
> 108
> │   exception =
> 109
> │    failed process: Process(`git push -fq upstream HEAD:gh-pages`, ProcessExited(1)) [1]
> 110
> │    
> 111
> │    Stacktrace:
> 112
> │      [1] pipeline_error
> ```
> 
> It is a git push command error, maybe a permission issue? I am not sure.
> 
> > (I've always just used the README to document modules.)
> 
> Me too. Actually this is a very good practice to host documents in README as long as a single page can do the job. Without documentation, one can still check the APIs in an REPL. So, no worry if the documentation issue can not be fixed.

I think @charleskawczynski started setting this up for me. I may remove the line from the README that gives badges for documentation. @charleskawczynski : Do you want to set this up? Ok by me if you simply want to let it all slide.

**Comment by charleskawczynski** [comment]
> has the documentation been successfully deployed before?

> I may remove the line from the README that gives badges for documentation. @charleskawczynski : Do you want to set this up?

No, it hasn't. There is still an issue with deployment. I'm able to merge PRs, but I don't have access to the settings, so I can't really help configure the deployment keys.

> Actually this is a very good practice to host documents in README as long as a single page can do the job. Without documentation, one can still check the APIs in an REPL. So, no worry if the documentation issue can not be fixed.

I think that's a good point, I was on the fence with suggesting to add the doc pages to start with, I think the docs are relatively short and just a readme is fine.

**Pull request merged**
